tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,223 @@
+# Stateful tests for arbitrary Zarr stores.
+
+import asyncio
+
+import hypothesis.strategies as st
+from hypothesis import note
+from hypothesis.stateful import (
+    Bundle,
+    RuleBasedStateMachine,
+    initialize,
+    invariant,
+    precondition,
+    rule,
+)
+
+from zarr.buffer import Buffer, default_buffer_prototype
+from zarr.store import MemoryStore
+
+# from strategies_store import StoreStatefulStrategies, key_ranges
+from zarr.testing.strategies import key_ranges, paths
+
+# zarr spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html
+
+
+class SyncStoreWrapper:
+    def __init__(self, store):
+        """Class to hold sync functions that map to async methods of MemoryStore
+        MemoryStore methods are async, this class' methods are sync, so just need to call asyncio.run() in them
+        then, methods in statemachine class are sync and call sync.
+        Unfortunately, hypothesis' stateful testing infra does not support asyncio
+        So we redefine sync versions of the Store API.
+        https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+        """
+        self.store = store
+        self.mode = store.mode
+
+    def set(self, key, data_buffer):  # buffer is value
+        return asyncio.run(self.store.set(key, data_buffer))
+
+    def list(self):
+        async def wrapper(gen):
+            return [i async for i in gen]
+
+        gen = self.store.list()  # async store list
+        return (i for i in asyncio.run(wrapper(gen)))
+
+    def get(self, key):
+        obs = asyncio.run(self.store.get(key, prototype=default_buffer_prototype()))
+        return obs
+
+    def get_partial_values(self, key_ranges):
```

Comment:
Let's preserve the original type signature: i.e. pass `prototype` in.

Reply:
I'm not sure how to handle this; if `prototype` is in the signature, then it would need to be passed to all of the rules that use `get()`, but one of the invariants also uses `get()` so this doesn't seem like a good option. I tried setting `prototype` as a default argument but this fails on precommit due to this error: 
![image](https://github.com/user-attachments/assets/adbcba69-7b05-4cf0-b70e-901f23167730)
The way i got it to work was by setting `default_buffer_prototype()` as an instance attr of the statemachine with `self.prototype = default_buffer_prototype()`, but this doesn't seem great, and it's still hardcoded since we can't pass a `prototype` arg to statemachine without switching to using `run_state_machine_as_test()` , any ideas on the best way to proceed?

Reply:
> but this doesn't seem great, and it's still hardcoded

This is OK. We can address later when someone needs this